### PR TITLE
Shear calculation added

### PR DESCRIPTION
**Pull request by SebereX**
Added shear calculation, checked to original MATLAB code. Have checked against some VMEC files (of preciseQA, preciseQH, +magwell); overall good agreement, except preciseQH+magWell,

### CODE REVIEW

**Comment by landreman** [comment]
The integration looks good, thanks.

Here are a couple more ideas for unit tests: 1) Compare to the shear calculation Wrick did for the case of circular x-section. 2) Compare to a case in axisymmetry (with nonzero I2 so the equations are well behaved).

The fact that integSig is periodic only for stellarator symmetry smells funny. Does that mean in the non-stellarator-symmetric case that the final integral around line 292 would be different if it were over all field periods instead of just over 1 field period? Can this periodic-only-in-stellarator-symmetry behavior be seen and understood also in Wrick's calculation for circular x-section (even if sigma(0)=0 the axis shape could be non-stellarator-symmetric) or in the case of axisymmetry?

**Comment by SebereX** [comment]
> The integration looks good, thanks.
> 
> Here are a couple more ideas for unit tests: 1) Compare to the shear calculation Wrick did for the case of circular x-section. 2) Compare to a case in axisymmetry (with nonzero I2 so the equations are well behaved).
I'll talk to Wrick about the possible comparison to the circular cross-section and I'll see if I concoct an axisymmetric comparison.
> 
> The fact that integSig is periodic only for stellarator symmetry smells funny. Does that mean in the non-stellarator-symmetric case that the final integral around line 292 would be different if it were over all field periods instead of just over 1 field period? Can this periodic-only-in-stellarator-symmetry behavior be seen and understood also in Wrick's calculation for circular x-section (even if sigma(0)=0 the axis shape could be non-stellarator-symmetric) or in the case of axisymmetry?

The expression for the shear, which is constructed by evaluating the ratio of two integrals in the exponential of `integSig`, will be the same regardless of the domain of integration (i.e., whether the whole domain or a single field period are considered). One may see this noting that the integrals in each of the field periods are proportional to each other (just a constant exponential factor to be factored out). Because these appear in numerator and denominator, the expression for shear is well described.

**Comment by SebereX** [comment]
@landreman  On the unit test: I have a reduced set of expressions for the AS case, which one can use to compare the more complicated general expressions and check that no obvious error exists (I have done that). However, this is still part of the NAE framework in some sense. It would be nice to check with respect to something from the outside (some VMEC run for example), but that proves tricky. You think the AS-NAE comparison is of any use? At least it checks things partially.

**Comment by landreman** [comment]
> @landreman On the unit test: I have a reduced set of expressions for the AS case, which one can use to compare the more complicated general expressions and check that no obvious error exists (I have done that). However, this is still part of the NAE framework in some sense. It would be nice to check with respect to something from the outside (some VMEC run for example), but that proves tricky. You think the AS-NAE comparison is of any use? At least it checks things partially.

Yes, I think a AS-NAE comparison would be a good test. It at least makes sure the new subroutine runs without crashing, for one thing, and the phi integrals in AS can be done analytically so it checks the varphi integration.

About nfp and `integSig`: Suppose the phi grid covered the full interval [0, 2pi) instead of just 1 field period. In the first field period, integSig(0)=0. But when stellarator symmetry is broken, `integSig` at the start of the 2nd field period would not be zero, due to the nonperiodic term `avSig*self.varphi`, right? Therefore `expSig` would be different in the 2nd field period than in the 1st, right?

**Comment by SebereX** [comment]
> > @landreman On the unit test: I have a reduced set of expressions for the AS case, which one can use to compare the more complicated general expressions and check that no obvious error exists (I have done that). However, this is still part of the NAE framework in some sense. It would be nice to check with respect to something from the outside (some VMEC run for example), but that proves tricky. You think the AS-NAE comparison is of any use? At least it checks things partially.
> 
> Yes, I think a AS-NAE comparison would be a good test. It at least makes sure the new subroutine runs without crashing, for one thing, and the phi integrals in AS can be done analytically so it checks the varphi integration.

I think I may only have the stellarator symmetric version of the AS limit (I need to check). Where shall I include the testing? Is it enough to include a couple of test cases to check? I am not used to these unit tests.
> 
> About nfp and `integSig`: Suppose the phi grid covered the full interval [0, 2pi) instead of just 1 field period. In the first field period, integSig(0)=0. But when stellarator symmetry is broken, `integSig` at the start of the 2nd field period would not be zero, due to the nonperiodic term `avSig*self.varphi`, right? Therefore `expSig` would be different in the 2nd field period than in the 1st, right?

Yes and no. Yes, because as you point `integSig` will have that non-periodic behaviour, but no because this has no effect on the computation of shear or any other physically relevant quantity. In the case of shear, because we are dividing by an integral also involving `integSig`, these factors you mention can be factored out and will cancel with the factors from the numerator. So at the end of the day it does not matter.

File: qsc/calculate_r3.py [hunk 1]
```diff
@@ -140,3 +143,159 @@ def calculate_r3(self):
...
+        2*X2s*Y2c + X31c*Y1s + 2*X2c*Y2s + X1c*Y31s) + curvature*(-2*X2s*Z2c + 2*X2c*Z2s + X1c*Z31s)) -
+        X31s*dXc1v - 2*X2s*dX2cdp + 2*X2c*dX2sdp + X1c*dX31sdp - Y31s*dY1cdp - 2*Y2s*dY2cdp +
+        2*Y2c*dY2sdp + Y1c*dY31sdp - 2*Z2s*dZ2cdp + 2*Z2c*dZ2sdp)
+
+    # Need to compute the integration factor necessary for computing the shear
+    DMred = d_d_varphi[1:,1:]   # The differentiation matrix has a linearly dependent row, focus on submatrix
+
+    # Distinguish between the stellarator symmetric case self.sigma0 = 0 and the non-symmetric ones.
+    # Distinction leads to the expSig function being periodic (stell. sym.) or not.
+    if self.sigma0 == 0:
```
Comment (landreman) [inline qsc/calculate_r3.py:279]: Should this line also check to see if the axis shape is non-stellarator-symmetric? (rs or zc nonzero)

**Review by landreman — COMMENTED** [review]
> I think I may only have the stellarator symmetric version of the AS limit (I need to check). Where shall I include the testing? Is it enough to include a couple of test cases to check? I am not used to these unit tests.

The latest commit has a new file `qsc/tests/test_iota2.py` with a couple of basic sanity tests. The comparison to iota2 in axisymmetry can go in there.

File: qsc/calculate_r3.py (hunk 1 above)
Comment (SebereX) [reply qsc/calculate_r3.py:279]: Yep. I was assuming that to be the case. Is there any specific property of the class specifying the symmetry, or do we need to check whether `rs` and `zc` are empty (or zeroes)?

Comment (landreman) [reply qsc/calculate_r3.py]: No worries, this line is now taken care of in the latest commit. Just wanted to check with you first. There is an `lasym` attribute, but it also accounts for B2s, whereas B2s doesn't affect symmetry of sigma.

**Comment by landreman** [comment]
Since it is turning out to be tricky to get a verification of the shear against vmec, let's go ahead and merge this PR now, and we can add such a test later. There are at least tests in place to run the `calculate_shear` subroutine.

**Pull request merged**